src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+---
+title: 'Understanding the ECMAScript spec, part 1'
+author: '[Marja Hölttä](https://twitter.com/marjakh), speculative specification spectator'
+avatars:
+  - marja-holtta
+date: 2020-01-22 13:33:37
+tags:
+  - ECMAScript
+description: 'Tutorial on reading the ECMAScript specification'
+tweet: ''
+---
+
+... where we take a simple function in the spec and try to understand the notation.
+
+## Preface
+
+Even if you know JavaScript, reading its language specification, the ECMAScript spec, can be pretty daunting. At least that's how I felt when I started reading it for the first time. Reading the spec from the start doesn't help either, as the spec is not designed to be read in that order, and notations are not explained before they occur.
+
+Let's start with a concrete example and walk through the spec to understand it. The following code demonstrates usage of Object.prototype.hasOwnProperty:
+
+```javascript
+let o = {'foo': 1};
+o.hasOwnProperty('foo'); // true
+o.hasOwnProperty('bar'); // false
+```
+
+In the example, _o_ doesn't have a property called _hasOwnProperty_, so we walk up the prototype chain and look for it. We find it in _o_'s prototype, which is _Object.prototype_.
+
+When trying to find out how _Object.prototype.hasOwnProperty_ is defined, I bumped into pseudocode-like descriptions like this:
+
+> Object.prototype.hasOwnProperty(V)
+>
+> When the hasOwnProperty method is called with argument V, the following steps are taken:
+>
+> 1. Let P be **?** ToPropertyKey(V).
+> 2. Let O be **?** ToObject(this value).
+> 3. Return **?** HasOwnProperty(O, P).
+
+[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)
+
+and
+
+> HasOwnProperty(O, P)
+>
+>The **abstract operation** HasOwnProperty is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments O and P where O is the object and P is the property key. This abstract operation performs the following steps:
+>
+> 1. **Assert**: Type(O) is Object.
+> 2. **Assert**: IsPropertyKey(P) is true.
+> 3. Let desc be **?** O.**\[\[GetOwnProperty\]\]**(P).
```

Comment:
Let's wrap these things in code spans, which both looks nicer and lets us avoid the weird escaping syntax:

```
> 1. **Assert**: `Type(O)` is `Object`.
> 2. **Assert**: `IsPropertyKey(P)` is `true`.
> 3. Let desc be `? O.[[GetOwnProperty]](P)`.
```

Reply:
I'm trying to maintain a style where quotes from the spec are inside >. Did you mean the code block should also be inside > (is that even possible?) or that the > should end before the code starts (confusing!)?

Reply:
Ah I read your comment again, I see what you mean. So not wrapping the whole thing inside a code block but individually. I might argue thought that the whole lines should be inside the code span, like. `Let desc be ...`

Reply:
Experimented a bit.

Option 1:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> `1. Let P be ? ToPropertyKey(V).`
> `2. Let O be ? ToObject(this value).`
> `3. Return ? HasOwnProperty(O, P).`

-> I think this looks ugly

Option 2:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> 1. Let `P` be ? `ToPropertyKey(V)`.
> 2. Let `O` be ? `ToObject(this value)`.
> 3. Return ? `HasOwnProperty(O, P)`.

-> I think this is confusing, also it's hard to decide what should be inside the code spans and what shouldn't. Let? Return? Question marks?

Option 3:

> When the hasOwnProperty method is called with argument V, the following steps are taken:
>
> 1. Let P be **?** ToPropertyKey(V).
> 2. Let O be **?** ToObject(this value).
> 3. Return **?** HasOwnProperty(O, P).

This is what I have now, the pros are that it allows using emphasis inside these blocks and it matches what the spec does. I still like this version best...

Reply:
Haha, I think I painted myself in a corner here :) I started changing "spec code snippets" inside the text to be inside `code spans`, and I think it's a good idea. But now they look different inside the text and inside the spec quotes. Argh.

Reply:
I consistently use option 2, with the difference that `?` is wrapped in the code span as well. It’s not an ordinary part of an English sentence, so it should be marked accordingly. IMHO this is the only semantically correct way of representing this content in Markdown (keeping in mind what HTML it produces).

I agree that option 1 is overkill (and ugly). Option 3 is… not great, semantically speaking.

Reply:
I'm convinced, going with option 2.